src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+Internal slots and methods are not accessible from JavaScript, for example, you cannot access _o.\[\[Prototype\]\]_ or call _o.\[\[GetOwnProperty\]\]()_. A JavaScript engine can implement them for their own internal use, but doesn't have to.
+
+Each internal method delegates to a similarly-named abstract operation:
+
+> \[\[GetOwnProperty\]\](P)
+>
+> When the \[\[GetOwnProperty\]\] internal method of O is called with property key P, the following steps are taken:
+>
+> Return ! OrdinaryGetOwnProperty(O, P).
+
+[Spec: \[\[GetOwnProperty\]\]](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-getownproperty-p)
+
+(We'll find out what the exclamation mark means in the next chapter.)
+
+_OrdinaryGetOwnProperty_ is not an internal method, since it's not associated with any object; instead, the object it operates on is passed as a parameter.
+
+_OrdinaryGetOwnProperty_ is called "ordinary" since it operates on ordinary objects. ECMAScript objects can be either **ordinary** or **exotic**. Ordinary objects must have the default behavior for a set of methods called **essential internal methods**. If an object deviates from the default behavior, it's exotic.
+
+The most well-known exotic object is the _Array_, since its length property behaves in a non-default way: setting the length property can remove elements from the _Array_.
+
+Essential internal methods are the methods listed [here](https://tc39.es/ecma262/#table-5).
+
+## Completion Records
+
+What about the question marks and exclamation marks? To understand them, we need to look into **Completion Records**!
+
+A Completion Record is a "record" (or a "struct", if you're more familiar with that term) defined by the spec. A record is a data type which has a fixed set of named fields. C-like languages have structs - JavaScript doesn't, since you can always add properties into an object.
+
+Again, the Completion Record is defined only for spec purposes. A JavaScript engine doesn't have to have a corresponding internal data type.
+
+A Completion Record has three fields:
+
+| Name | Description |
+--- | ---
+| \[\[Type\]\] | One of: normal, break, continue, return, or throw. All other types except "normal" are referred to as "abrupt completions".|
+| \[\[Value\]\] | the value that was produced when the completion occurred, e.g., the return value of a function or the exception (if one is thrown)|
+| \[\[Target\]\] | used for directed control transfers (ignoring it for this blog post)|
+
+[Spec: Completion Record](https://tc39.es/ecma262/#sec-completion-record-specification-type)
+
+Every abstract operation implicitly returns a Completion Record. Even if it looks like an abstract operation would return a simple type such as Boolean, it's implicitly wrapped into a Completion Record with the type "normal" (see [Implicit Completion Values](https://www.ecma-international.org/ecma-262/index.html#sec-implicit-completion-values)).
```

Comment:
We're trying to change this! The implicitness doesn't work in all contexts and is generally confusing. Maybe a note that the editors are currently exploring making all completion record handling more explicit.